### PR TITLE
options: Add a static parse API

### PR DESCRIPTION
**Pull request by mwestphal**

### CODE REVIEW

File: library/src/options.cxx [hunk 1]
```diff
@@ -105,6 +105,48 @@ std::pair<std::string, unsigned int> options::getClosestOption(const std::string
...
+  return options_tools::parse<std::string>(str);
+}
+
+//----------------------------------------------------------------------------
+template<>
+std::vector<double> options::parse(const std::string& str)
+{
+  return options_tools::parse<std::vector<double>>(str);
+}
+
```
Comment (Meakk) [inline library/src/options.cxx:149]: what about doing
```cpp
template<typename T>
T options::parse(const std::string& str)
{
  return options_tools::parse<T>(str);
}

template bool options::parse<bool>(const std::string& str);
template int options::parse<int>(const std::string& str);
// ...
```

that way, you have a oneliner when you need to add a type.

Comment (Meakk) [reply library/src/options.cxx:149]: You could even do something like:
```cpp
#define F3D_DECL_TYPE(TYPE) template TYPE options::parse<TYPE>(const std::string& str)
F3D_DECL_TYPE(bool);
F3D_DECL_TYPE(int);
// ...
```
Yes I know, macros are bad but in this case it makes sense.

Comment (mwestphal) [reply library/src/options.cxx]: fair point.

File: library/src/options.cxx [hunk 2]
```diff
@@ -105,6 +105,22 @@ std::pair<std::string, unsigned int> options::getClosestOption(const std::string
...
+//----------------------------------------------------------------------------
+#define F3D_DECL_TYPE(TYPE) template TYPE options::parse<TYPE>(const std::string& str)
+F3D_DECL_TYPE(bool);
+F3D_DECL_TYPE(int);
+F3D_DECL_TYPE(double);
+F3D_DECL_TYPE(f3d::ratio_t);
+F3D_DECL_TYPE(std::string);
+F3D_DECL_TYPE(std::vector<double>);
```
Comment (Meakk) [inline library/src/options.cxx:122]:
```suggestion
//----------------------------------------------------------------------------
template<typename T>
T options::parse(const std::string& str)
{
  return options_tools::parse<T>(str);
}

//----------------------------------------------------------------------------
#define F3D_DECL_TYPE(TYPE) template F3D_EXPORT TYPE options::parse<TYPE>(const std::string& str)
F3D_DECL_TYPE(bool);
F3D_DECL_TYPE(int);
F3D_DECL_TYPE(double);
F3D_DECL_TYPE(f3d::ratio_t);
F3D_DECL_TYPE(std::string);
F3D_DECL_TYPE(std::vector<double>);
```

**Review by Meakk — CHANGES_REQUESTED** [review]
This fix the Windows build issue

File: library/public/options.h.in [hunk 3]
```diff
@@ -102,6 +106,14 @@ public:
    */
   std::pair<std::string, unsigned int> getClosestOption(const std::string& option) const;
 
+  /**
+   * Templated parsing method used internally to parse strings.
+   * Implemented only for: bool, int, double, ratio_t, string and double vector.
+   * Throw an options::parsing_exception if parsing failed.
+   */
+  template<typename T>
+  static F3D_EXPORT T parse(const std::string& str);
```
Comment (Meakk) [inline library/public/options.h.in:115]: Actually, I'm wondering if this `F3D_EXPORT` is required now.

**Review by Meakk — APPROVED** [review]

**Pull request merged**